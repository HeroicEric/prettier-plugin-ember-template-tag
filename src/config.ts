// FIXME: Get ETI to export these + types
// @ts-expect-error
import * as util from 'ember-template-imports/src/util';

export const TEMPLATE_TAG_NAME: string = util.TEMPLATE_TAG_NAME;
export const TEMPLATE_TAG_PLACEHOLDER: string = util.TEMPLATE_TAG_PLACEHOLDER;

export const TEMPLATE_TAG_OPEN = `<${TEMPLATE_TAG_NAME}>`;
export const TEMPLATE_TAG_CLOSE = `</${TEMPLATE_TAG_NAME}>`;
