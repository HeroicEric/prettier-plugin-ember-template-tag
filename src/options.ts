import type {
  BooleanSupportOption,
  ParserOptions,
  SupportOptions,
} from 'prettier';

import type { BaseNode } from './types/ast';

export interface Options extends ParserOptions<BaseNode> {
  templateExportDefault?: boolean;
  templateSingleQuote?: boolean;
}

const templateExportDefault: BooleanSupportOption = {
  since: '0.1.0',
  category: 'Format',
  type: 'boolean',
  default: false,
  description: 'Prepend default export template tags with "export default".',
};

/**
 * Extracts a valid `templateSingleQuote` option out of the provided options. If
 * `templateSingleQuote` is defined, it will be used, otherwise the value for
 * `singleQuote` will be inherited.
 */
export function getTemplateSingleQuote(options: Options): boolean {
  const { singleQuote, templateSingleQuote } = options;
  return typeof templateSingleQuote === 'boolean'
    ? templateSingleQuote
    : singleQuote;
}

const templateSingleQuote: BooleanSupportOption = {
  since: '0.0.3',
  category: 'Format',
  type: 'boolean',
  description:
    'Use single quotes instead of double quotes within template tags.',
};

export const options: SupportOptions = {
  templateExportDefault,
  templateSingleQuote,
};
