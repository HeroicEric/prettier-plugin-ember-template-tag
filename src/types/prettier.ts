import type { doc } from 'prettier';
import { isRecord } from '../utils/index';

/** Type predicate */
export function isDocGroup(value: unknown): value is doc.builders.Group {
  return isRecord(value) && value['type'] === 'group';
}
