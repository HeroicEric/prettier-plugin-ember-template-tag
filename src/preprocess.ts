import { preprocessEmbeddedTemplates } from 'ember-template-imports/lib/preprocess-embedded-templates';
import type { BaseNode } from 'estree';
import type { Parser } from 'prettier';

import { TEMPLATE_TAG_NAME, TEMPLATE_TAG_PLACEHOLDER } from './config';

export const preprocess: Required<Parser<BaseNode>>['preprocess'] = (
  text,
  options
) => {
  return preprocessEmbeddedTemplates(text, {
    // FIXME: Need to figure this out for ESLint but probably OK not to do it
    // for prettier support?
    getTemplateLocals(_template: string): string[] {
      return [];
    },

    templateTag: TEMPLATE_TAG_NAME,
    templateTagReplacement: TEMPLATE_TAG_PLACEHOLDER,

    includeSourceMaps: false,
    includeTemplateTokens: false,

    relativePath: options.filepath
  }).output;
};
