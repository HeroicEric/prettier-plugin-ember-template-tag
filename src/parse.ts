import { traverse } from '@babel/core';
import type { Node } from '@babel/types';
import { type Comment } from '@babel/types';
import { Preprocessor } from 'content-tag';
import type { Parser } from 'prettier';
import { parsers as babelParsers } from 'prettier/plugins/babel.js';

import { PRINTER_NAME } from './config';
import type { Options } from './options.js';
import type { GlimmerTemplate } from './types/glimmer';
import { assert } from './utils';

const typescript = babelParsers['babel-ts'] as Parser<Node | undefined>;
const p = new Preprocessor();

interface Prepared {
  output: string;
  templateInfos: {
    type: 'expression' | 'class-member';
    tagName: 'template';
    contents: string;
    range: {
      start: number;
      end: number;
    };
    contentRange: {
      start: number;
      end: number;
    };
    startRange: {
      end: number;
      start: number;
    };
    endRange: {
      start: number;
      end: number;
    };
  }[];
}

interface PreprocessedResult {
  /** Range of the template including <template></template> tags */
  templateRange: readonly [start: number, end: number];
  /** Range of the template content, excluding <template></template> tags */
  range: readonly [start: number, end: number];
  ast: GlimmerTemplate;
}

/** Traverses the AST and replaces the transformed template parts with other AST */
function convertAst(
  result: { ast: Node; code: string },
  preprocessedResult: PreprocessedResult[],
): void {
  // FIXME:
  const templateInfos = preprocessedResult;
  let counter = 0;

  traverse(result.ast, {
    enter(path) {
      const node = path.node;
      if (
        node.type === 'ObjectExpression' ||
        node.type === 'BlockStatement' ||
        node.type === 'StaticBlock'
      ) {
        const range = node.range as [number, number];

        const template = templateInfos.find(
          (t) =>
            (t.templateRange[0] === range[0] &&
              t.templateRange[1] === range[1]) ||
            (t.templateRange[0] === range[0] - 1 &&
              t.templateRange[1] === range[1] + 1) ||
            (t.templateRange[0] === range[0] &&
              t.templateRange[1] === range[1] + 1),
        );

        if (!template) {
          return null;
        }
        counter++;
        assert('expected ast on template', template.ast);
        const { ast } = template;
        ast.extra.isAlreadyExportDefault =
          path.parent.type === 'ExportDefaultDeclaration' ||
          path.parentPath?.parent.type === 'ExportDefaultDeclaration';
        ast.extra.isDefaultTemplate =
          path.parent.type === 'ExportDefaultDeclaration' ||
          path.parent.type === 'Program' ||
          (path.parent.type === 'ExpressionStatement' &&
            path.parentPath?.parent.type === 'Program') ||
          (path.parent.type === 'TSAsExpression' &&
            path.parentPath?.parentPath?.parent.type === 'Program') ||
          path.parentPath?.parent.type === 'ExportDefaultDeclaration';

        ast.extra.isAssignment =
          !ast.extra.isDefaultTemplate && node.type !== 'StaticBlock';

        ast.leadingComments = node.leadingComments as Comment[];
        Object.assign(node, ast);
      }
      return null;
    },
  });

  if (counter !== templateInfos.length) {
    throw new Error('failed to process all templates');
  }
}

/**
 * Preprocesses the template info, parsing the template content to Glimmer AST,
 * fixing the offsets and locations of all nodes also calculates the block
 * params locations & ranges and adding it to the info
 */
function preprocess(info: Prepared, code: string): PreprocessedResult[] {
  return info.templateInfos.map((tpl) => {
    const range = [tpl.contentRange.start, tpl.contentRange.end] as const;
    const templateRange = [tpl.range.start, tpl.range.end] as const;
    const template = code.slice(...range);
    return {
      templateRange,
      range,
      ast: {
        type: 'FunctionDeclaration',
        leadingComments: [],
        range: [templateRange[0], templateRange[1]],
        start: templateRange[0],
        end: templateRange[1],
        extra: {
          isGlimmerTemplate: true,
          isDefaultTemplate: false,
          isAssignment: false,
          isAlreadyExportDefault: false,
          template,
        },
      },
    };
  });
}

function replaceRange(
  s: string,
  start: number,
  end: number,
  substitute: string,
): string {
  return s.slice(0, start) + substitute + s.slice(end);
}

function prepare(code: string): Prepared {
  let jsCode = code;
  const result = p.parse(code) as Prepared['templateInfos'];
  for (const tplInfo of result.reverse()) {
    const lineBreaks = [...tplInfo.contents].reduce(
      (previous, current) => previous + (current === '\n' ? 1 : 0),
      0,
    );
    if (tplInfo.type === 'class-member') {
      const tplLength = tplInfo.range.end - tplInfo.range.start;
      const spaces = tplLength - 'static{`'.length - '`}'.length - lineBreaks;
      const total = ' '.repeat(spaces) + '\n'.repeat(lineBreaks);
      const replacementCode = `static{\`${total}\`}`;
      jsCode = replaceRange(
        jsCode,
        tplInfo.range.start,
        tplInfo.range.end,
        replacementCode,
      );
    } else {
      const tplLength = tplInfo.range.end - tplInfo.range.start;
      const nextWord = code.slice(tplInfo.range.end).match(/\S+/);
      let prefix = '{';
      let suffix = '}';
      if (nextWord && nextWord[0] === 'as') {
        prefix = '(' + prefix;
        suffix = suffix + ')';
      } else if (!nextWord || ![',', ')'].includes(nextWord[0][0] || '')) {
        suffix += ';';
      }
      const spaces = tplLength - prefix.length - suffix.length - lineBreaks;
      const total = ' '.repeat(spaces) + '\n'.repeat(lineBreaks);
      const replacementCode = `${prefix}${total}${suffix}`;
      jsCode = replaceRange(
        jsCode,
        tplInfo.range.start,
        tplInfo.range.end,
        replacementCode,
      );
    }
  }
  return {
    templateInfos: result,
    output: jsCode,
  };
}

export const parser: Parser<Node | undefined> = {
  ...typescript,
  astFormat: PRINTER_NAME,

  async parse(code: string, options: Options): Promise<Node> {
    const prepared = prepare(code);
    const preprocessedResult = preprocess(prepared, code);
    const ast = await typescript.parse(prepared.output, options);
    assert('expected ast', ast);
    convertAst({ ast, code }, preprocessedResult);
    return ast;
  },
};
