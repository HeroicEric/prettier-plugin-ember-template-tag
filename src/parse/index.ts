import { traverse } from '@babel/core';
import type { Comment, Node } from '@babel/types';
import { Preprocessor } from 'content-tag';
import type { Parser } from 'prettier';
import { parsers as babelParsers } from 'prettier/plugins/babel.js';

import { PRINTER_NAME } from '../config';
import type { Options } from '../options.js';
import type { GlimmerTemplate, RawGlimmerTemplate } from '../types/glimmer';
import { isDefaultTemplate } from '../types/glimmer';
import { assert } from '../utils';
import { normalizeWhitespace } from './whitespace';

const typescript = babelParsers['babel-ts'] as Parser<Node | undefined>;
const p = new Preprocessor();

interface Preprocessed {
  code: string;
  results: PreprocessedResult[];
}

interface PreprocessedResult {
  /**
   * Range of the contents, inclusive of inclusive of the
   * `<template></template>` tags.
   */
  range: readonly [start: number, end: number];

  /**
   * Range of the template contents, not inclusive of the
   * `<template></template>` tags.
   */
  contentRange: readonly [start: number, end: number];

  templateNode: GlimmerTemplate;
}

/** Traverses the AST and replaces the transformed template parts with other AST */
function convertAst(ast: Node, preprocessedResult: PreprocessedResult[]): void {
  let counter = 0;

  traverse(ast, {
    enter(path) {
      const { node } = path;
      if (
        node.type === 'ObjectExpression' ||
        node.type === 'BlockStatement' ||
        node.type === 'StaticBlock'
      ) {
        const { range } = node;
        assert('expected range', range);

        const preprocessedTemplate = preprocessedResult.find(
          (p) =>
            (p.range[0] === range[0] && p.range[1] === range[1]) ||
            (p.range[0] === range[0] - 1 && p.range[1] === range[1] + 1) ||
            (p.range[0] === range[0] && p.range[1] === range[1] + 1),
        );

        if (!preprocessedTemplate) {
          return null;
        }

        const { templateNode } = preprocessedTemplate;
        templateNode.extra.isDefaultTemplate = isDefaultTemplate(path);

        templateNode.leadingComments = node.leadingComments as Comment[];

        Object.assign(node, templateNode);

        counter++;
      }
      return null;
    },
  });

  if (counter !== preprocessedResult.length) {
    throw new Error('failed to process all templates');
  }
}

/**
 * Pre-processes the template info, parsing the template content to Glimmer AST,
 * fixing the offsets and locations of all nodes also calculates the block
 * params locations & ranges and adding it to the info
 */
function preprocess(code: string): Preprocessed {
  const templateNodes = p.parse(code) as RawGlimmerTemplate[];
  const results: PreprocessedResult[] = [];
  let output = code;
  for (const templateNode of templateNodes.reverse()) {
    output = normalizeWhitespace(templateNode, code, output);

    const contentRange = [
      templateNode.contentRange.start,
      templateNode.contentRange.end,
    ] as const;
    const range = [templateNode.range.start, templateNode.range.end] as const;
    const template = code.slice(...contentRange);
    const ast: GlimmerTemplate = {
      type: 'FunctionDeclaration',
      leadingComments: [],
      range: [range[0], range[1]],
      start: range[0],
      end: range[1],
      extra: {
        isGlimmerTemplate: true,
        isDefaultTemplate: false,
        template,
      },
    };
    results.push({
      range,
      contentRange,
      templateNode: ast,
    } satisfies PreprocessedResult);
  }

  return { results, code: output };
}

export const parser: Parser<Node | undefined> = {
  ...typescript,
  astFormat: PRINTER_NAME,

  async parse(code: string, options: Options): Promise<Node> {
    const preprocessed = preprocess(code);
    const ast = await typescript.parse(preprocessed.code, options);
    assert('expected ast', ast);
    convertAst(ast, preprocessed.results);
    return ast;
  },
};
