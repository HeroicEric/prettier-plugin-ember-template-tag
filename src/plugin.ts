import { doc, Parser } from 'prettier';
const { builders } = doc;

const languages = [
  {
    extensions: ['.gjs'],
    name: 'Ember Template Tag',
    parsers: ['gjs-parse']
  }
];

const parsers: Record<string, Parser<unknown>> = {
  'gjs-parse': {
    // call the preprocessor here? (or in preprocess below)
    // then use babel or estree to parse it
    // replace the glimmer nodes w/ custom nodes, then register parsers for the custom nodes (see the angular parser)
    parse: text => parser.parse(text),
    astFormat: 'estree'

    // parse: (text: string, parsers: { [parserName: string]: Parser }, options: ParserOptions<T>) => T;
    // astFormat: string;
    // hasPragma?: ((text: string) => boolean) | undefined;
    // locStart: (node: T) => number;
    // locEnd: (node: T) => number;
    // preprocess?: ((text: string, options: ParserOptions<T>) => string) | undefined;
    // https://github.com/prettier/prettier/blob/main/src/language-js/loc.js
  }
};

function printToml(path, options, print) {
  const node = path.getValue();

  if (Array.isArray(node)) {
    return concat(path.map(print));
  }

  switch (node.type) {
    default:
      return '';
  }
}

const printers = {
  'toml-ast': {
    print: printToml
  }
};

module.exports = {
  languages,
  parsers,
  printers
};
