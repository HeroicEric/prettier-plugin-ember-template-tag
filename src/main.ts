import type { BaseNode } from 'estree';
import type { Parser, ParserOptions, Printer, SupportLanguage } from 'prettier';
import { parsers as babelParsers } from 'prettier/parser-babel';

import { PARSER_NAME, PRINTER_NAME } from './config';
import { preprocess } from './preprocess';
import { definePrinter, printer } from './print/index';

const typescript = babelParsers['babel-ts'] as Parser<BaseNode>;

export const languages: SupportLanguage[] = [
  {
    extensions: ['.gjs', '.gts'],
    name: 'Ember Template Tag',
    parsers: [PARSER_NAME]
  }
];

export const parsers: Record<string, Parser<BaseNode>> = {
  [PARSER_NAME]: {
    ...typescript,
    astFormat: PRINTER_NAME,

    preprocess(text: string, options: ParserOptions<BaseNode>): string {
      definePrinter(options);
      let js = preprocess(text, options);
      return typescript.preprocess?.(js, options) ?? js;
    }
  }
};

export const printers: Record<string, Printer<BaseNode>> = {
  [PRINTER_NAME]: printer
};
