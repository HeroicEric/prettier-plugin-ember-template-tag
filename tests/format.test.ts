import fs from 'fs';
import path from 'path';
import { format, Options } from 'prettier';
import { describe, expect, test } from 'vitest';

import { PARSER_NAME } from '../src/config';
import plugin from '../src/main';

const formatOptions: Options = {
  parser: PARSER_NAME,
  plugins: [plugin]
};

describe('format', () => {
  const caseDir = path.join(__dirname, './cases');
  const cases = fs.readdirSync(caseDir, { withFileTypes: true }).map(file => {
    const name = file.name;
    const filename = path.join(caseDir, name);
    const text = fs.readFileSync(filename, 'utf8');

    return {
      name: file.name,
      filename,
      code: text
    };
  });

  for (let testCase of cases) {
    test(`it formats ${testCase.name}`, () => {
      const code = `
        <template>
                what
              </template>
      `;
      let result = format(code, formatOptions);
      expect(result).toMatchSnapshot();
    });
  }
});
