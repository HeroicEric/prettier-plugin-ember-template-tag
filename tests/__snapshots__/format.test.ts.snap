// Vitest Snapshot v1

exports[`format > it formats simple.gjs 1`] = `
"<template>what</template>
"
`;
